### PR TITLE
ecosystems and dpkg coalescer

### PR DESCRIPTION
**Pull request by ldelossa**
This PR introduces the concept of "Ecosystem" which acts are a coupling of similar scanning/coalescing tool chains. This replaces arbitrary factories.

PR also implements a DPKG coalescer which retrieves discovered packages and distributions the dpkg/os-release scanner found in several layers and computes a final image result.

### CODE REVIEW

File: internal/scanner/layerscanner/layerscanner.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+// Scan performs a concurrency controlled scan of each layer by each type of configured scanner; indexing
+// the results on successful completion.
+//
+// Scan will launch all necessary go routines and each routine will block on adding a token.
+// On completion a token is discarded unblocking other routines which are waiting.
+//
+// On ctx cancel or a go routine reporting an scan/index error all routines blocking on adding a token will error
+// and the will not subsequently try to discard a token.
```
Comment (hdonnay) [inline internal/scanner/layerscanner/layerscanner.go]: Again, what are tokens? This should probably start "On cancellation or the first reported non-nil error,". There's also something missing here.

**Comment by ldelossa** [comment]
@hdonnay think token bucket is the wrong model, thats for rate limiting. This is a channel driven semaphore i believe. 

Any suggestions on improving ?

File: internal/scanner/layerscanner/layerscanner.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	case <-ls.cc:
+	default:
+	}
+}
+
+// Scan performs a concurrency controlled scan of each layer by each type of configured scanner; indexing
+// the results on successful completion.
+//
+// Scan will launch all necessary go routines and each routine will block on adding a token.
+// On completion a token is discarded unblocking other routines which are waiting.
```
[the remark this answers is not in the document]
Comment (ldelossa) [reply internal/scanner/layerscanner/layerscanner.go]: The "addToken" and "discardToken" are not part of the interface? Figured it was self explanatory from those methods.  Token being the empty meaningless struct

[the remark this answers is not in the document]
Comment (hdonnay) [reply internal/scanner/layerscanner/layerscanner.go]: Yes, this is documentation on an exported method talking about unexported implementation details.

**Pull request merged**